riscv-runtime/src/io.rs

```diff
@@ -7,6 +9,94 @@ use powdr_riscv_syscalls::Syscall;
 use alloc::vec;
 use alloc::vec::Vec;
 
+/// An iterator over the static prover data.
+#[derive(Copy, Clone)]
+pub struct ProverDataReader {
+    remaining_data: &'static [u32],
+}
+
+use spin::Mutex;
+
+static PROVER_DATA_READER: Mutex<Option<ProverDataReader>> = Mutex::new(None);
+
+// Initialize the reader once and get a mutable reference
+fn get_prover_data_reader() -> spin::MutexGuard<'static, Option<ProverDataReader>> {
+    let mut reader = PROVER_DATA_READER.lock();
+    if reader.is_none() {
+        *reader = Some(ProverDataReader::new());
+    }
+    reader
+}
+
+impl ProverDataReader {
+    /// Creates an iterator over the static prover data.
+    ///
+    /// A newly created iterator will start at the beginning of the prover data.
+    pub fn new() -> Self {
+        extern "C" {
+            // The prover data start and end symbols. Their addresses are set by the linker.
+            static __powdr_prover_data_start: u32;
+            static __powdr_prover_data_end: u32;
+        }
+        const POWDR_PAGE_SIZE: isize = 2048;
+
+        unsafe {
+            // We skip the first page of the prover data, as it used as salt to
+            // randomize its merkle tree node.
+            let region_start: *const u32 = &__powdr_prover_data_start;
+            let data_start = region_start.byte_offset(POWDR_PAGE_SIZE);
+            let data_end: *const u32 = &__powdr_prover_data_end;
+
+            let prover_data_section =
+                slice::from_raw_parts(data_start, data_end.offset_from(data_start) as usize);
```

Comment:
i think only this statement is unsafe no?

Reply:
No, `byte_offset` is also unsafe.

The unsafe block could encompass just the two lines, but when I wrote I felt that taking the pointers and trimming the slice also fit the feng shui of the unsafe block.